File: src/check.ts
import { createPublicClient, http, parseAbi, type PublicClient } from "viem";
import { mainnet } from "viem/chains";
import { isStale } from "./isStale.js";
import { quoteFromFeed } from "./quote.js";
import { lookupFeed } from "./feeds.js";
import { SEQUENCER_FEEDS } from "./feeds.js";
import { checkSequencer } from "./sequencer.js";

/**
 * Official Data Feed ABI — only `decimals` and `latestRoundData` per
 * https://docs.chain.link/data-feeds/api-reference
 */
const feedAbi = parseAbi([
  "function decimals() view returns (uint8)",
  "function latestRoundData() view returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)",
] as const);

export type CheckPricesInput = {
  rpc: string;
  feeds: Array<{
    feed: string;
    maxAgeSeconds: number;
    amountEth?: number | null;
  }>;
  nowSeconds?: number;
  /** @internal inject mock client for tests */
  __client?: Pick<PublicClient, "readContract" | "multicall"> & {
    getChainId?: PublicClient["getChainId"];
  };
};

export type CheckPriceInput = {
  rpc: string;
  feed: string;
  maxAgeSeconds: number;
  amountEth?: number | null;
  nowSeconds?: number;
  /** @internal inject mock client for tests (no live RPC) — viem-compatible */
  __client?: Pick<PublicClient, "readContract"> & { getChainId?: PublicClient["getChainId"] };
};

export type CheckPriceResult = {
  decision: "ALLOW" | "BLOCK";
  reason: string;
  feed: string;
  answer: string;
  priceUsd: number | null;
  amountEth: number | null;
  quoteUsd: number | null;
  updatedAt: string;
  ageSeconds: number | null;
  maxAgeSeconds: number;
  now: number;
  allowExecute: boolean;
};

function blockResult(
  input: CheckPriceInput,
  reason: string,
  partial?: Partial<CheckPriceResult>,
): CheckPriceResult {
  const now = input.nowSeconds ?? Math.floor(Date.now() / 1000);
  return {
    decision: "BLOCK",
    reason,
    feed: input.feed,
    answer: partial?.answer ?? "0",
    priceUsd: partial?.priceUsd ?? null,
    amountEth: input.amountEth ?? null,
    quoteUsd: partial?.quoteUsd ?? null,
    updatedAt: partial?.updatedAt ?? "0",
    ageSeconds: partial?.ageSeconds ?? null,
    maxAgeSeconds: input.maxAgeSeconds,
    now,
    allowExecute: false,
  };
}

/**
 * Full check: viem `latestRoundData` + `decimals` → `isStale` → `quoteFromFeed`.
 * Fail closed on any RPC/read/zero/negative/invalid. No wallet, no write.
 * Verified against viem `createPublicClient` + `readContract` docs.
 */

/**
 * Check multiple price feeds in a single RPC multicall.
 * This is heavily optimized for high-frequency agents and rebalancing operations.
 * Fetches latestRoundData, decimals, and sequencer uptime simultaneously.
 */
export async function checkPrices(input: CheckPricesInput): Promise<CheckPriceResult[]> {
  const { rpc, feeds } = input;
  const now = input.nowSeconds ?? Math.floor(Date.now() / 1000);

  // Return empty immediately if no feeds
  if (feeds.length === 0) return [];

  const client =
    input.__client ??
    createPublicClient({
      chain: mainnet,
      transport: http(rpc),
    });

  // 1. Initial validations and registry lookups
  type PreparedFeed = {
    originalIndex: number;
    feed: string;
    maxAgeSeconds: number;
    amountEth: number | null;
    entry: ReturnType<typeof lookupFeed>;
    blockedResult?: CheckPriceResult;
  };

  const prepared: PreparedFeed[] = feeds.map((f, i) => {
    const amountEth = f.amountEth ?? null;
    const baseBlock = {
      rpc,
      feed: f.feed,
      maxAgeSeconds: f.maxAgeSeconds,
      amountEth,
      nowSeconds: now,
      __client: input.__client,
    };

    if (amountEth !== null) {
      if (typeof amountEth !== "number" || !Number.isFinite(amountEth) || amountEth < 0) {
        return {
          originalIndex: i,
          feed: f.feed,
          maxAgeSeconds: f.maxAgeSeconds,
          amountEth,
          entry: null,
          blockedResult: blockResult(
            baseBlock,
            `invalid amountEth ${String(amountEth)} — BLOCK (fail closed)`,
          ),
        };
      }
    }
    if (!/^0x[a-fA-F0-9]{40}$/.test(f.feed)) {
      return {
        originalIndex: i,
        feed: f.feed,
        maxAgeSeconds: f.maxAgeSeconds,
        amountEth,
        entry: null,
        blockedResult: blockResult(baseBlock, `invalid feed ${f.feed} — BLOCK`),
      };
    }
    const entry = lookupFeed(f.feed);
    if (!entry) {
      return {
        originalIndex: i,
        feed: f.feed,
        maxAgeSeconds: f.maxAgeSeconds,
        amountEth,
        entry: null,
        blockedResult: blockResult(baseBlock, `unknown/unsupported feed ${f.feed} — BLOCK`),
      };
    }
    if (
      !Number.isFinite(f.maxAgeSeconds) ||
      !Number.isInteger(f.maxAgeSeconds) ||
      f.maxAgeSeconds < 0
    ) {
      return {
        originalIndex: i,
        feed: f.feed,
        maxAgeSeconds: f.maxAgeSeconds,
        amountEth,
        entry,
        blockedResult: blockResult(
          baseBlock,
          `invalid maxAgeSeconds ${String(f.maxAgeSeconds)} — BLOCK`,
        ),
      };
    }
    if (typeof rpc !== "string" || rpc.trim() === "") {
      return {
        originalIndex: i,
        feed: f.feed,
        maxAgeSeconds: f.maxAgeSeconds,
        amountEth,
        entry,
        blockedResult: blockResult(baseBlock, "missing rpc — BLOCK (fail closed)"),
      };
    }
    return { originalIndex: i, feed: f.feed, maxAgeSeconds: f.maxAgeSeconds, amountEth, entry };
  });

  // Verify all valid feeds belong to the same chain to ensure the RPC matches
  const validFeeds = prepared.filter((p) => !p.blockedResult && p.entry);
  if (validFeeds.length > 0) {
    const targetChainId = validFeeds[0].entry!.chainId;
    for (const p of validFeeds) {
      if (p.entry!.chainId !== targetChainId) {
        p.blockedResult = blockResult(
          {
            rpc,
            feed: p.feed,
            maxAgeSeconds: p.maxAgeSeconds,
            amountEth: p.amountEth,
            nowSeconds: now,
            __client: input.__client,
          },
          `chain mismatch in batch: expected ${targetChainId}, got ${p.entry!.chainId} — BLOCK`,
        );
      }
    }

    // Check RPC chainId
    const maybeGetChainId = (client as unknown as { getChainId?: () => Promise<number> })
      .getChainId;
    if (typeof maybeGetChainId === "function") {
      try {
        const rpcChainId = await maybeGetChainId.call(client);
        for (const p of validFeeds) {
          if (!p.blockedResult && p.entry!.chainId !== rpcChainId) {
            p.blockedResult = blockResult(
              {
                rpc,
                feed: p.feed,
                maxAgeSeconds: p.maxAgeSeconds,
                amountEth: p.amountEth,
                nowSeconds: now,
                __client: input.__client,
              },
              `chainId mismatch: feed ${p.feed} is chainId ${p.entry!.chainId} but rpc returned chainId ${String(rpcChainId)} — BLOCK`,
            );
          }
        }
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        for (const p of validFeeds) {
          if (!p.blockedResult)
            p.blockedResult = blockResult(
              {
                rpc,
                feed: p.feed,
                maxAgeSeconds: p.maxAgeSeconds,
                amountEth: p.amountEth,
                nowSeconds: now,
                __client: input.__client,
              },
              `failed to get chainId — BLOCK (fail closed): ${msg}`,
            );
        }
      }
    }
  }

  const finalResults: CheckPriceResult[] = new Array(feeds.length);
  const stillValid = prepared.filter((p) => !p.blockedResult);

  // If no feeds are still valid, return the blocked ones
  if (stillValid.length === 0) {
    for (const p of prepared) {
      finalResults[p.originalIndex] = p.blockedResult!;
    }
    return finalResults;
  }

  // 2. Prepare Multicall
  const targetChainId = stillValid[0].entry!.chainId;
  // We'll optionally add sequencer call at the end of the contracts array if needed
  const hasSequencer = SEQUENCER_FEEDS[targetChainId] !== undefined;

  const contracts = [];
  for (const p of stillValid) {
    contracts.push({
      address: p.feed as `0x${string}`,
      abi: feedAbi,
      functionName: "latestRoundData",
    });
    contracts.push({ address: p.feed as `0x${string}`, abi: feedAbi, functionName: "decimals" });
  }
  if (hasSequencer) {
    contracts.push({
      address: SEQUENCER_FEEDS[targetChainId] as `0x${string}`,
      abi: feedAbi,
      functionName: "latestRoundData",
    });
  }

  // Execute Multicall
  let multicallResults: any[] = [];
  try {
    multicallResults = await client.multicall({
      contracts,
      allowFailure: true,
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    // Multicall completely failed
    for (const p of stillValid) {
      p.blockedResult = blockResult(
        {
          rpc,
          feed: p.feed,
          maxAgeSeconds: p.maxAgeSeconds,
          amountEth: p.amountEth,
          nowSeconds: now,
          __client: input.__client,
        },
        `multicall failed — BLOCK (fail closed): ${msg}`,
      );
    }
  }

  // Evaluate Sequencer
  let seqErrorMsg: string | null = null;
  if (hasSequencer && multicallResults.length > 0) {
    const seqRes = multicallResults[multicallResults.length - 1];
    if (seqRes.status === "failure") {
      seqErrorMsg = `failed to read sequencer feed on chain ${targetChainId} — BLOCK (fail closed)`;
    } else {
      const seqRoundData = seqRes.result as readonly [bigint, bigint, bigint, bigint, bigint];
      const seqAnswer = seqRoundData[1];
      const seqStartedAt = seqRoundData[2];
      if (seqAnswer === 1n) {
        seqErrorMsg = `L2 Sequencer is DOWN on chain ${targetChainId} — BLOCK`;
      } else if (seqAnswer === 0n) {
        const timeSinceUp = now - Number(seqStartedAt);
        const GRACE_PERIOD = 3600;
        if (timeSinceUp < GRACE_PERIOD) {
          seqErrorMsg = `L2 Sequencer is in grace period (${timeSinceUp}s < ${GRACE_PERIOD}s) on chain ${targetChainId} — BLOCK`;
        }
      }
    }
  }

  // Evaluate individual feeds
  if (multicallResults.length > 0) {
    for (let i = 0; i < stillValid.length; i++) {
      const p = stillValid[i];
      if (p.blockedResult) continue;

      if (seqErrorMsg) {
        p.blockedResult = blockResult(
          {
            rpc,
            feed: p.feed,
            maxAgeSeconds: p.maxAgeSeconds,
            amountEth: p.amountEth,
            nowSeconds: now,
            __client: input.__client,
          },
          seqErrorMsg,
        );
        continue;
      }

      const lrdRes = multicallResults[i * 2];
      const decRes = multicallResults[i * 2 + 1];

      if (lrdRes.status === "failure") {
        p.blockedResult = blockResult(
          {
            rpc,
            feed: p.feed,
            maxAgeSeconds: p.maxAgeSeconds,
            amountEth: p.amountEth,
            nowSeconds: now,
            __client: input.__client,
          },
          `failed to read latestRoundData from ${p.feed} — BLOCK (fail closed)`,
        );
        continue;
      }
      if (decRes.status === "failure") {
        p.blockedResult = blockResult(
          {
            rpc,
            feed: p.feed,
            maxAgeSeconds: p.maxAgeSeconds,
            amountEth: p.amountEth,
            nowSeconds: now,
            __client: input.__client,
          },
          `failed to read decimals() from ${p.feed} — BLOCK (fail closed)`,
        );
        continue;
      }

      const data = lrdRes.result as readonly [bigint, bigint, bigint, bigint, bigint];
      const roundId = data[0];
      const answer = data[1];
      const updatedAt = data[3];
      const answeredInRound = data[4];

      const d = decRes.result;
      const decimals = typeof d === "bigint" ? Number(d) : Number(d);

      const baseBlock = {
        rpc,
        feed: p.feed,
        maxAgeSeconds: p.maxAgeSeconds,
        amountEth: p.amountEth,
        nowSeconds: now,
        __client: input.__client,
      };
      if (!Number.isInteger(decimals) || decimals < 0 || decimals > 36) {
        p.blockedResult = blockResult(
          baseBlock,
          `failed to read decimals() from ${p.feed} — BLOCK (fail closed): invalid decimals ${String(d)}`,
        );
        continue;
      }
      if (updatedAt === 0n) {
        p.blockedResult = blockResult(
          baseBlock,
          `updatedAt is 0 (no data) from ${p.feed} — BLOCK`,
          { answer: answer.toString(), updatedAt: updatedAt.toString(), ageSeconds: null },
        );
        continue;
      }
      if (answer <= 0n) {
        p.blockedResult = blockResult(
          baseBlock,
          `answer is ${answer.toString()} (invalid price) from ${p.feed} — BLOCK`,
          { answer: answer.toString(), updatedAt: updatedAt.toString(), ageSeconds: null },
        );
        continue;
      }
      if (answeredInRound < roundId) {
        p.blockedResult = blockResult(
          baseBlock,
          `incomplete round: answeredInRound ${answeredInRound.toString()} < roundId ${roundId.toString()} (unanswered round) from ${p.feed} — BLOCK`,
          { answer: answer.toString(), updatedAt: updatedAt.toString(), ageSeconds: null },
        );
        continue;
      }

      const stale = isStale({ updatedAt, nowSeconds: now, maxAgeSeconds: p.maxAgeSeconds });

      let priceUsd: number | null = null;
      let quoteUsd: number | null = null;
      try {
        const q = quoteFromFeed({ answer, decimals, amountEth: p.amountEth });
        priceUsd = q.priceUsd;
        quoteUsd = q.quoteUsd;
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        p.blockedResult = {
          decision: "BLOCK",
          reason: `quote failed: ${msg} — BLOCK`,
          feed: p.feed,
          answer: answer.toString(),
          priceUsd: null,
          amountEth: p.amountEth,
          quoteUsd: null,
          updatedAt: updatedAt.toString(),
          ageSeconds: stale.ageSeconds,
          maxAgeSeconds: p.maxAgeSeconds,
          now,
          allowExecute: false,
        };
        continue;
      }

      p.blockedResult = {
        decision: stale.decision,
        reason: stale.reason,
        feed: p.feed,
        answer: answer.toString(),
        priceUsd,
        amountEth: p.amountEth,
        quoteUsd,
        updatedAt: updatedAt.toString(),
        ageSeconds: stale.ageSeconds,
        maxAgeSeconds: p.maxAgeSeconds,
        now,
        allowExecute: stale.decision === "ALLOW",
      };
    }
  }

  // Populate final array
  for (const p of prepared) {
    finalResults[p.originalIndex] = p.blockedResult!;
  }

  return finalResults;
}

/**
 * Single price check. Convenience wrapper around checkPrices.
 */
export async function checkPrice(input: CheckPriceInput): Promise<CheckPriceResult> {
  const res = await checkPrices({
    rpc: input.rpc,
    feeds: [
      {
        feed: input.feed,
        maxAgeSeconds: input.maxAgeSeconds,
        amountEth: input.amountEth,
      },
    ],
    nowSeconds: input.nowSeconds,
    __client: input.__client as any,
  });
  return res[0];
}

File: src/check.rs
use crate::abi::{decode_round_data, decode_word_u128};
use crate::addressbook::is_valid_eth_address;
use crate::feeds::lookup_feed;
use crate::is_stale::{is_stale, IsStaleInput};
use crate::quote::{quote_from_feed, QuoteInput};
use crate::rpc::EvmRpcClient;
use crate::sequencer::check_sequencer;
use crate::types::Decision;
use serde::{Deserialize, Serialize};

pub const LATEST_ROUND_DATA_SELECTOR: &str = "0xfeaf968c";
pub const DECIMALS_SELECTOR: &str = "0x313ce567";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckPriceResult {
    pub decision: Decision,
    pub reason: String,
    pub feed: String,
    pub answer: String,
    pub price_usd: Option<f64>,
    pub amount_eth: Option<f64>,
    pub quote_usd: Option<f64>,
    pub updated_at: String,
    pub age_seconds: Option<i64>,
    pub max_age_seconds: i64,
    pub now: i64,
    pub allow_execute: bool,
}

#[derive(Debug, Clone)]
pub struct CheckPriceInput<'a> {
    pub feed: &'a str,
    pub max_age_seconds: i64,
    pub amount_eth: Option<f64>,
    pub now_seconds: Option<i64>,
}

fn block_result(
    feed: &str,
    reason: String,
    max_age_seconds: i64,
    now: i64,
    amount_eth: Option<f64>,
) -> CheckPriceResult {
    CheckPriceResult {
        decision: Decision::Block,
        reason,
        feed: feed.to_string(),
        answer: "0".to_string(),
        price_usd: None,
        amount_eth,
        quote_usd: None,
        updated_at: "0".to_string(),
        age_seconds: None,
        max_age_seconds,
        now,
        allow_execute: false,
    }
}

pub async fn check_price(
    client: &dyn EvmRpcClient,
    input: CheckPriceInput<'_>,
) -> CheckPriceResult {
    let now = input
        .now_seconds
        .unwrap_or_else(|| chrono::Utc::now().timestamp());

    if now < 0 {
        return block_result(
            input.feed,
            "invalid now timestamp (negative) — BLOCK".to_string(),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    if !is_valid_eth_address(input.feed) {
        return block_result(
            input.feed,
            format!("invalid feed address {} — BLOCK", input.feed),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    if input.max_age_seconds < 0 {
        return block_result(
            input.feed,
            "maxAgeSeconds must be >= 0 — BLOCK".to_string(),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    // Check if feed is allowlisted in registry
    let feed_entry = match lookup_feed(input.feed) {
        Some(e) => e,
        None => {
            return block_result(
                input.feed,
                format!("unknown feed {} (not allowlisted) — BLOCK", input.feed),
                input.max_age_seconds,
                now,
                input.amount_eth,
            );
        }
    };

    // If on an L2 chain, check sequencer status
    if let Some(seq_block) = check_sequencer(feed_entry.chain_id, client, now as u64).await {
        return block_result(
            input.feed,
            seq_block,
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    // Query round data and decimals
    let (round_res, dec_res) = tokio::join!(
        client.call(input.feed, LATEST_ROUND_DATA_SELECTOR),
        client.call(input.feed, DECIMALS_SELECTOR),
    );

    let hex_round = match round_res {
        Ok(h) => h,
        Err(e) => {
            return block_result(
                input.feed,
                format!(
                    "failed to read latestRoundData() from {} — BLOCK (fail closed): {}",
                    input.feed, e
                ),
                input.max_age_seconds,
                now,
                input.amount_eth,
            );
        }
    };

    let hex_dec = match dec_res {
        Ok(h) => h,
        Err(e) => {
            return block_result(
                input.feed,
                format!(
                    "failed to read decimals() from {} — BLOCK (fail closed): {}",
                    input.feed, e
                ),
                input.max_age_seconds,
                now,
                input.amount_eth,
            );
        }
    };

    let (round_id, answer, _started_at, updated_at, answered_in_round) =
        match decode_round_data(&hex_round) {
            Ok(d) => d,
            Err(e) => {
                return block_result(
                    input.feed,
                    format!(
                        "failed to decode latestRoundData() from {}: {}",
                        input.feed, e
                    ),
                    input.max_age_seconds,
                    now,
                    input.amount_eth,
                );
            }
        };

    let decimals = match decode_word_u128(&hex_dec, 0) {
        Ok(d) if d <= 18 => d as u8,
        _ => {
            return block_result(
                input.feed,
                format!("invalid decimals response from {}", input.feed),
                input.max_age_seconds,
                now,
                input.amount_eth,
            );
        }
    };

    if updated_at == 0 {
        return block_result(
            input.feed,
            format!("updatedAt is 0 (no data) from {} — BLOCK", input.feed),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    if answer <= 0 {
        return block_result(
            input.feed,
            format!(
                "answer is {} (invalid price) from {} — BLOCK",
                answer, input.feed
            ),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    if answered_in_round < round_id {
        return block_result(
            input.feed,
            format!(
                "incomplete round: answeredInRound {} < roundId {} (unanswered round) from {} — BLOCK",
                answered_in_round, round_id, input.feed
            ),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }

    // updatedAt is uint80 on-chain; reject values unrepresentable as i64
    // instead of silently wrapping via `as` cast.
    if updated_at > i64::MAX as u128 {
        return block_result(
            input.feed,
            format!(
                "updatedAt {} exceeds i64::MAX (unrepresentable) from {} — BLOCK",
                updated_at, input.feed
            ),
            input.max_age_seconds,
            now,
            input.amount_eth,
        );
    }
    let updated_at_i64 = updated_at as i64;

    let stale_check = is_stale(IsStaleInput {
        updated_at: Some(updated_at_i64),
        now_seconds: now,
        max_age_seconds: input.max_age_seconds,
    });

    let (price_usd, quote_usd) = match quote_from_feed(QuoteInput {
        answer,
        decimals,
        amount: input.amount_eth,
    }) {
        Ok(q) => (Some(q.price_usd), q.quote_usd),
        Err(e) => {
            return CheckPriceResult {
                decision: Decision::Block,
                reason: format!("quote failed: {} — BLOCK", e),
                feed: input.feed.to_string(),
                answer: answer.to_string(),
                price_usd: None,
                amount_eth: input.amount_eth,
                quote_usd: None,
                updated_at: updated_at.to_string(),
                age_seconds: stale_check.age_seconds,
                max_age_seconds: input.max_age_seconds,
                now,
                allow_execute: false,
            };
        }
    };

    CheckPriceResult {
        decision: stale_check.decision,
        reason: stale_check.reason,
        feed: input.feed.to_string(),
        answer: answer.to_string(),
        price_usd,
        amount_eth: input.amount_eth,
        quote_usd,
        updated_at: updated_at.to_string(),
        age_seconds: stale_check.age_seconds,
        max_age_seconds: input.max_age_seconds,
        now,
        allow_execute: stale_check.decision == Decision::Allow,
    }
}

pub async fn check_prices(
    client: &dyn EvmRpcClient,
    feeds: Vec<CheckPriceInput<'_>>,
) -> Vec<CheckPriceResult> {
    let mut results = Vec::new();
    for f in feeds {
        results.push(check_price(client, f).await);
    }
    results
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::feeds::DEFAULT_FEED;
    use crate::mock::MockRpcClient;
    use std::sync::Arc;

    #[tokio::test]
    async fn test_check_price_fresh_allowed() {
        let now = 1700000050;
        let round_data_hex = format!(
            "0x{:0>64x}{:0>64x}{:0>64x}{:0>64x}{:0>64x}",
            10u64, 2500_00000000u64, 1700000000u64, 1700000040u64, 10u64
        );
        let dec_hex = format!("0x{:0>64x}", 8u64);

        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, data| {
                if data == LATEST_ROUND_DATA_SELECTOR {
                    Ok(round_data_hex.clone())
                } else if data == DECIMALS_SELECTOR {
                    Ok(dec_hex.clone())
                } else {
                    Err("unknown".to_string())
                }
            })),
            ..Default::default()
        };

        let res = check_price(
            &mock,
            CheckPriceInput {
                feed: DEFAULT_FEED,
                max_age_seconds: 60,
                amount_eth: Some(1.0),
                now_seconds: Some(now),
            },
        )
        .await;

        assert_eq!(res.decision, Decision::Allow);
        assert!(res.allow_execute);
        assert!((res.price_usd.unwrap() - 2500.0).abs() < 1e-3);
        assert_eq!(res.age_seconds, Some(10));
    }

    #[tokio::test]
    async fn test_check_price_stale_blocked() {
        let now = 1700000200; // 160s old
        let round_data_hex = format!(
            "0x{:0>64x}{:0>64x}{:0>64x}{:0>64x}{:0>64x}",
            10u64, 2500_00000000u64, 1700000000u64, 1700000040u64, 10u64
        );
        let dec_hex = format!("0x{:0>64x}", 8u64);

        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, data| {
                if data == LATEST_ROUND_DATA_SELECTOR {
                    Ok(round_data_hex.clone())
                } else if data == DECIMALS_SELECTOR {
                    Ok(dec_hex.clone())
                } else {
                    Err("unknown".to_string())
                }
            })),
            ..Default::default()
        };

        let res = check_price(
            &mock,
            CheckPriceInput {
                feed: DEFAULT_FEED,
                max_age_seconds: 60,
                amount_eth: None,
                now_seconds: Some(now),
            },
        )
        .await;

        assert_eq!(res.decision, Decision::Block);
        assert!(!res.allow_execute);
        assert!(res.reason.contains("stale"));
    }

    #[tokio::test]
    async fn test_check_price_updated_at_overflow_blocked() {
        let now = 1700000200;
        let huge_updated_at: u128 = 1u128 << 70; // exceeds i64::MAX
        let round_data_hex = format!(
            "0x{:0>64x}{:0>64x}{:0>64x}{:0>64x}{:0>64x}",
            10u64, 2500_00000000u64, 1700000000u64, huge_updated_at, 10u64
        );
        let dec_hex = format!("0x{:0>64x}", 8u64);

        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, data| {
                if data == LATEST_ROUND_DATA_SELECTOR {
                    Ok(round_data_hex.clone())
                } else if data == DECIMALS_SELECTOR {
                    Ok(dec_hex.clone())
                } else {
                    Err("unknown".to_string())
                }
            })),
            ..Default::default()
        };

        let res = check_price(
            &mock,
            CheckPriceInput {
                feed: DEFAULT_FEED,
                max_age_seconds: 60,
                amount_eth: None,
                now_seconds: Some(now),
            },
        )
        .await;

        assert_eq!(res.decision, Decision::Block);
        assert!(!res.allow_execute);
        assert!(res.reason.contains("unrepresentable"));
    }

    #[tokio::test]
    async fn test_check_price_unknown_feed_blocked() {
        let mock = MockRpcClient::default();
        let res = check_price(
            &mock,
            CheckPriceInput {
                feed: "0x0000000000000000000000000000000000000000",
                max_age_seconds: 60,
                amount_eth: None,
                now_seconds: Some(1000),
            },
        )
        .await;

        assert_eq!(res.decision, Decision::Block);
        assert!(res.reason.contains("not allowlisted"));
    }
}
